### PR TITLE
Workshop3: Workshop Core

### PR DESCRIPTION
**Pull request by Solxanich**
### What is the new feature?
The implementation for re-adapting Mod Browser, Publishing Mods, Downloading Mods.

There are pretty-up things that can be done afterwards, and I know I'll miss something someplace with how many inter-mixed functionalities there are but I believe I've dealt with it all in one form or another.

Allows for:
1) Downloading & Updating Workshop Items
2) Removing installed Workshop Items
4) Publishing mods to Workshop
5) Browsing Mods on the Workshop

### Why should this be part of tModLoader?
This moves us from the legacy 1.3 Mod Browser to the new Steam Workshop Mod Browser. Its a fairly big step, and a lot more pretty-ness & features can be added hereafter.

### Are there alternative designs?
Willing to modify as needed to clean-up code

Known Outstanding Concerns/ Areas for future improvements:
- [x] Mod Browser refreshes too early, and doesn't catch the item has been downloaded
- [ ] Can't seem to re-download a deleted item within the same session - this issue lies on Steam's end, unfortunately, from testing done.
- [ ] CommandLine & Github integrations for publishing
- [x] https://github.com/tModLoader/tModLoader/issues/1569
- [ ] Verifying GoG and China downloading for tModLoader (GoG verified against Terraria, prelim China results from Wallpaper Engine suggests 2 in 3 chance it will work)
- [ ] How Workshop Collections work within our Workshop
- [x] Memory Usage during querying - some concern to how much memory it may need at a 1000 mods or more

### CODE REVIEW

File: patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModBrowser.cs [hunk 1]
```diff
@@ -202,217 +192,93 @@ internal partial class UIModBrowser : UIState, IHaveBackButtonCommand
...
-		//		private void OnModDownloadCompleted(HttpDownloadRequest req) {
-		//			if (req.Response.StatusCode != HttpStatusCode.OK) {
-		//				string errorKey = req.Response.StatusCode == HttpStatusCode.ServiceUnavailable ? "MBExceededBandwidth" : "MBUnknownMBError";
-		//				Interface.errorMessage.Show(Language.GetTextValue("tModLoader." + errorKey), 0);
-		//			}
-		//			else if (req.Success && req.CustomData is UIModDownloadItem currentDownload) {
-		//				ProcessDownloadedMod(req, currentDownload);
-		//			}
-		//		}
+		internal UIModDownloadItem FindModDownloadItem(string publishId)
```
Comment (Solxanich) [inline patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModBrowser.cs:274]: Revert this, and instead just box PublishId dependency within Workshop implement, not browser.
Mods should be based on traditional ModReferences by modname

File: patches/tModLoader/Terraria/ModLoader/UI/Interface.cs [hunk 2]
```diff
@@ -339,25 +329,16 @@ internal static class Interface
...
-							else {
-								string tempFile = ModLoader.ModPath + Path.DirectorySeparatorChar + "temporaryDownload" + DownloadFile.TEMP_EXTENSION;
-								client.DownloadFile(downloadURL, tempFile);
-
-								if (ModLoader.TryGetMod(modname, out var mod))
-									mod.Close();
-
-								File.Copy(tempFile, ModLoader.ModPath + Path.DirectorySeparatorChar + downloadURL.Substring(downloadURL.LastIndexOf("/")), true);
-								File.Delete(tempFile);
+						if (modBrowser.Items.Count == 0) {
```
Comment (Solxanich) [inline patches/tModLoader/Terraria/ModLoader/UI/Interface.cs:332]: Replace with WorkshopHelper.QueryHelper.CheckWorkshopConnection

**Review by Solxanich — COMMENTED** [review]
Some improvements needed:
1) Cleanup Checking Workshop Connection to be a helper function
2) Cleanup Workshop Metadata tags to be less hardcoded
3) Devert unnecessary change for ModReferences going from ModName to PublishID. ModReferences is used elsewhere outside of workshop context.

**Review by Mirsario — APPROVED** [review]

**Comment by Solxanich** [comment]
GoG Verified: 

https://user-images.githubusercontent.com/59670736/126093792-4aedde3c-ac44-428e-b401-d2526d5227bd.mp4

**Pull request merged**